### PR TITLE
Add lint to warn for calls to `std::mem::drop` with a reference argument

### PR DESCRIPTION
**Pull request by inrustwetrust**
This is a possible footgun, especially when a mutex lock guard is involved.

### CODE REVIEW

File: src/drop_ref.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        lint_array!(DROP_REF)
+    }
+}
+
+impl LateLintPass for DropRefPass {
+    fn check_expr(&mut self, cx: &LateContext, expr: &Expr) {
+        if let ExprCall(ref path, ref args) = expr.node {
+            if let ExprPath(None, _) = path.node {
+                let def_id = cx.tcx.def_map.borrow()[&path.id].def_id();
+                if match_def_path(cx, def_id, &["core", "mem", "drop"]) {
```
Comment (Manishearth) [inline src/drop_ref.rs]: This should be a constant path along with the others in utils.rs

File: src/drop_ref.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    }
+}
+
+fn check_drop_arg(cx: &LateContext, call_span: Span, arg: &Expr) {
+    let arg_ty = cx.tcx.expr_ty(arg);
+    if let ty::TyRef(..) = arg_ty.sty {
+        span_note_and_lint(cx,
+                           DROP_REF,
+                           call_span,
+                           "call to `std::mem::drop` with a reference argument",
```
Comment (Manishearth) [inline src/drop_ref.rs]: you can mention here that dropping a reference does nothing.

File: src/drop_ref.rs (hunk 1 above)
Comment (inrustwetrust) [inline src/drop_ref.rs]: Ah, missed that this was the convention. Will fix.

**Comment by inrustwetrust** [comment]
Addressed review feedback.

**Comment by llogiq** [comment]
The update_lints.py check failed, but the code looks good now. You can either re-run `util/update_lints.py` and commit the `README.md` or I can do it while merging. :+1:

**Comment by inrustwetrust** [comment]
Just saw it as well. Have rebased on latest trunk now, this should fix it I hope.

**Comment by llogiq** [comment]
Yes it did. Thank you for your contribution! :smile:

**Comment by inrustwetrust** [comment]
Thanks for a great project :+1:

**Pull request merged**